src/forward/compiler.jl

```diff
@@ -70,7 +70,91 @@ function dual(ir)
   return ir
 end
 
-@dynamo function _pushforward(_, x...)
+const chainrules_frule_fallback = which(frule, Tuple{Any})
+
+"""
+  has_chain_frule(T)
+
+For a type-tuple `T` e.g. `Tuple{typeof(f), Int, Float64}`, checks if there is a `rrule` defined for it.
+Excluding the generic fallback.
+The first return value is `true` if the `rrule` exists, `false` otherwise.
+If it does not, then the second argument is a list of edges to attach to the CodeInfo for a generated function,
+such that if a suitable rule is defined later, the generated function will recompile.
+"""
+function has_chain_frule(T)
+  m = meta(Tuple{typeof(frule), T.parameters...})
+
+  if m.method !== chainrules_frule_fallback
+    # found a frule, no need to add any edges
+    return true, nothing
+  end
+
+  return false, m.instance
+end
+
+"""
+    chain_frule(f, args...)
+
+Returns a the (primal) value of `f(args...)` and tangent, by invoking
+`ChainRules.frule(f, args...)`.
+"""
+@noinline chain_frule(dargs, args...) = frule(dargs, args...)
+
+"""
+  chain_frule_kw(kwf, kwargs, f, args...)
+
+As per [`chain_frule`](@ref) but with support for kwargs.
+`kwf` should be the kwfunc matching to `f`, and `kwargs` are a `NamedTuple` of keyword
+arguments.
+"""
+@inline chain_frule_kw(kwf, kwargs, f, args...) = frule(f, args...; kwargs...)
+
+ignore_sig(T) = all(T -> T <: Type, T.parameters)
+
+"""
+    is_kwfunc(sigt...)
+
+Determines if `sigt` is the type signature of a kwfunction.
+Each element of `sigt` should be a type.
+Either the first 3 types are a kwfunc type, a NamedTuple and the matching base function type,
+or the first argument is the base function type and it is not a kwfunction.
+the remaining types in `sigt` are the types of the argument.
+
+"""
+is_kwfunc(::Vararg{Any}) = false
+is_kwfunc(k, ::Type{<:NamedTuple}, f, args...) = k===Core.kwftype(f)
```

Comment:
this code exists already for reverse mode we should just import it from there?

Reply:
Yes. I'll need to refactor to avoid all of this duplication. I just copied + pasted to start with because I was tired haha